### ISSUE TITLE
Can't use Chrome debugger in Visual Studio Code

### ISSUE THREAD

**Issue by screendriver** [fuse-box/fuse-box#659]
I don't know why but I can't use [Debugger for Chrome](https://marketplace.visualstudio.com/items?itemName=msjsdiag.debugger-for-chrome) when I use a fuse-box build (SourceMaps are enabled). Strangely enough I can see and debug my TypeScript code in Chrome DevTools, but I can't set breakpoints in Visual Studio Code.

Does it require a different kind of SourceMaps than fuse-box generates?

**Comment by nchanged** [comment]
@screendriver I think it does require different sourcemaps per file.
https://stackoverflow.com/questions/31169259/how-to-debug-typescript-files-in-visual-studio-code

**Comment by screendriver** [comment]
Hmm I don't know if this is a project setting at my project or it's a task that `fuse-box` should do. For example with webpack this works at my machine out of the box. I don't have to change anything in my project 🤔 

In my `tsconfig.json` I only can enable `sourceMap` or `inlineSourceMap`. In `fuse-box` it's the same: only `sourceMap: true` can be set in `FuseBox.init()`. That's all I can do by my side.

**Comment by andrew-w-ross** [comment]
@screendriver I've made a [fuse-debug](https://github.com/andrew-w-ross/fuse-debug) repo as an example on how to get it working.

The change that got it working is [webroot](https://github.com/andrew-w-ross/fuse-debug/blob/master/.vscode/launch.json#L9). If you are having trouble determining how the vscode is attempting to map the source-maps type in `.scripts` whilst debugging in the debug panel.

@nchanged With this all being said I think there is a bug with sourceMaps.  Following the documentation [here](http://fuse-box.org/page/configuration#sourcemaps), I can't seem to get the `sourceRoot` or the `inline` option to work.

You can have a look at [my attempt](https://github.com/andrew-w-ross/fuse-debug/blob/master/fuse.js#L6). If I set the option from `true` to `{ sourceRoot: "/src", inline: false }` no source-maps seem to be generated at all. What I am expecting to happen is an almost identical source-map with the exception `"sourceRoot":"/src"` and `"sourcesContent":[null,null,null]`.

**Comment by auchenberg** [comment]
Referencing, https://github.com/Microsoft/vscode-chrome-debug/issues/459

**Comment by nchanged** [comment]
@auchenberg any ideas how to fix sourcemaps with HMR? it's runtime. Therefore sourcemaps cannot be mapped anymore.

**Comment by auchenberg** [comment]
@nchanged I've opened a seperat issue which tracks the sourcemapping problem in fuse-box.https://github.com/fuse-box/fuse-box/issues/673#issuecomment-315698239

**Comment by nchanged** [comment]
any updates on the issue?

**Comment by andrew-w-ross** [comment]
@nchanged Quite a bit has happened while researching this issue so I'll try to summaries the story so far.
To clarify the source-maps fail on update. Initial investigation found that the update was incorrectly pointing to the original file path map file, that obviously doesn't exist, not the bundle map file so when it tried to find the o. 

I corrected this [in my branch ](https://github.com/andrew-w-ross/fuse-box/tree/fix-hmr-sourcemap) but no dice. It still couldn't navigate to the original map file, the problem is the [chrome event-scriptParsed](https://chromedevtools.github.io/devtools-protocol/tot/Debugger/#event-scriptParsed) event. 

That url argument is how vscode determines where to download the .map file. The problem is chrome does not set it when using the function constructor, which is how fuse-box does it's updates.

My last attempt was then to just hard code the `sourceURL` to the absolute path of the bundle output path and hopefully that would work again no dice. [This is the last time myselft](Microsoft/vscode-chrome-debug/issues/459#issuecomment-317955023) and @auchenberg spoke about it. 

Again the last time I tried to fix this was in July so things might work.

tl;dr: Fuse-box's source mapping fails on update as it's broken but on fixing revealed vscode can't handle the use case where function construction is used to do the patch.

The quick fix would be to add patch scripts links to the head of the document the same way webpack  does. With that being said I far prefer the current approach fuse-box takes to patching.

**Comment by nchanged** [comment]
Could you create a repository with instructions on how to reproduce the issue, I might have some ideas

**Comment by andrew-w-ross** [comment]
Here are the steps to recreate it:

- Check out my fixed version of fuse-box `git clone -b fix-hmr-sourcemap https://github.com/andrew-w-ross/fuse-box.git ` 
- Do the install, build and link:
  - `cd fuse-box`
  - `npm i`
  - `npm run dist`
  - `npm link`
- Then check out my debug project `git clone -b fix-hmr-sourcemap https://github.com/andrew-w-ross/fuse-debug.git ` 
- Do the install and link:
  - `cd fuse-debug`
  - `npm i`
  - `npm link fuse-box`
- open the project with vs code and launch with the `Start Both` configuration. 
- add a break point to [the add function](https://github.com/andrew-w-ross/fuse-debug/blob/fix-hmr-sourcemap/src/components/Adder.tsx#L19)
- click the click me button and and the break point should hit
- update the value in Adder to some other number and save, wait for the page to hot reload and the breakpoint should go from red -> grey -> red again
- click the click me button and the breakpoint won't hit but the value updates correctly.

**Comment by nchanged** [comment]
### FuseBox v4 Major Upgrade

As of today, the latest major version of FuseBox `v4` is available in `fuse-box@next`. FuseBox has undergone a full re-write from scratch, learning from its mistakes in the previous versions.

There are a lot of improvements, speed and config wise.

Get started [here](https://github.com/fuse-box/fuse-box/blob/master/docs/getting-started/get-started.md)

We are still working on the documentation and polishing before the major release. We are getting very close. And we need help testing.

```bash
npm install fuse-box@next
```

Most of the issues that I am closing have been resolved, however, if you find that your issue is still relevant in `v4` don't hesitate to re-open or (better) create a new one with `v4` configuration.
